Introduce the next_*day and prev_*day helpers for DateAndTime related Classes

andela-oakinniranye opened the pull request:
### Added The prev__day and next__day methods to the DateAndTime related objects

Often when working on time related applications, such as scheduling applications, there is a need for helpers like next_tuesday, next_wednesday, next_sunday. While Rails provides the next_week(:tuesday), it may not often be what's required by the application developer. For instance, if today were a Thursday and the developer wanted to build a scheduling application that allows the user to select from a list Fridays, using Date#next_week(:friday), would result in a week being lost in the list, which may not be the required behaviour. Same applies for the Date#prev_week(:friday).
### Usage Example

With this implementation, to get the next 4 Thursdays for a list, it can easily be achieved with the snippet below:

``` ruby
    4.times{ |i| 
       Date.today.next_thursday + i.week
    }
```
### Methods Introduced

This PR adds the following methods to the DateAndTime related objects: 
`[:next_sunday, :next_monday, :next_tuesday, :next_wednesday, :next_thursday, :next_friday, :next_saturday, :prev_sunday, :prev_monday, :prev_tuesday, :prev_wednesday, :prev_thursday, :prev_friday, :prev_saturday]`
### Related Feature Request

This PR was built on the related Feature Request on https://groups.google.com/forum/?fromgroups#!topic/rubyonrails-core/suUoWFlOsPI

cc @fxn

Commit:
Introduce the next_*day and prev_*day helpers for DateAndTime related Classes

Often when working on time dependent applications, such as scheduling applications, there is a need for helpers like next_tuesday, next_wednesday, next_sunday. While Rails provides the next_week(:tuesday), it may not often be what's required by the application developer. For instance, if today were a Thursday and the developer wanted to build a scheduling application that allows the user to select from a list Fridays, using Date#next_week(:friday), would result in a week being lost in the list, which may not be the required behaviour. Same applies for the Date#prev_week(:friday).
However, this implementation would allow the developer to add such support as
    4.times{|i| Date.today.next_thursday + i.week}
This PR resolves this problem following a related Feature Request on https://groups.google.com/forum/?fromgroups#!topic/rubyonrails-core/suUoWFlOsPI
It adds the following methods to the DateAndTime related objects: [:next_sunday, :next_monday, :next_tuesday, :next_wednesday, :next_thursday, :next_friday, :next_saturday, :prev_sunday, :prev_monday, :prev_tuesday, :prev_wednesday, :prev_thursday, :prev_friday, :prev_saturday]

sgrif commented:
We usually don't accept new features to Active Support unless there is an overwhelmingly common use case that it solves. These methods are definitely useful, but they'll be adding to an already crowded namespace for all applications. I think this makes more sense as a standalone gem, rather than being included in Rails itself. I'll let @arthurnn or @fxn chime in if they disagree though

sgrif commented:
Ignore the last comment, I missed that this was accepted on the mailing list.

sgrif commented:
r? @fxn

Commit:
Previously I copied the comment for the next_*day methods, which instead of using the wording such as previous *day before self, it used next *day after self. That's been updated, with an additional CHANGE.log info on usage example and renaming a local variable to achieve more clarity

andela-oakinniranye commented:
Test failures are unrelated to the build!

fxn commented:
ACK! I'll check this one soon.

fxn commented:
Some feedback.

I think the implementation could be simplified a bit:

```ruby
DAYS_INTO_WEEK.each do |day_name, index|
  n = (index + 1) % 7 # To match Ruby's wday.

  define_method "next_#{day_name}" do
    days_advance  = n - wday
    days_advance += 7 if days_advance <= 0
    advance(days: days_advance)
  end

  define_method "prev_#{day_name}" do
    days_advance  = n - wday
    days_advance -= 7 if days_advance >= 0
    advance(days: days_advance)
  end
end
```

It's straightforward and symmetric.

Then, regarding the Changelog entry, let's try to make it a bit shorter. For starters, we can avoid talking about `next_week`. That method is really unrelated to this feature, it does not attempt to do this so no need to say it does not. Also, Rails documentation should not use `Date.today` generally speaking. In a Rails application the date is idiomatically *relative to the application time zone*, never to the system time zone. Better use `Date.current` instead.

There may be some minor edits, but I can do those myself no prob.

matthewd commented:
If we're going to add this, I think `next_day(:tuesday)` would be a better fit with the rest of the API.

jeremy commented:
I'd find this useful. Agree with treating the weekday (`Date#wday`) as the argument, not part of the method name. I'd expect it to work with numeric wdays, too, so we can `Date.today.next_day(last_occurred_on.wday)` without converting to symbols.

matthewd commented:
@jeremy in that case, perhaps we should co-opt `next_weekday` instead? `next_day(3)` sounds like something else.

jeremy commented:
Thought that too, but we already claimed next_weekday as "non-weekend days"
rather than "days of week." Could do next_wday.
On Sat, Jun 4, 2016 at 14:03 Matthew Draper <notifications@github.com>
wrote:

> @jeremy <https://github.com/jeremy> in that case, perhaps we should
> co-opt next_weekday instead? next_day(3) sounds like something else.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/rails/pull/25209#issuecomment-223778170>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAAAx-ZwoSAKJRmvO5z3646lf1js_v49ks5qIegEgaJpZM4IqdWw>
> .
>

fxn commented:
So, `next_day` for the name, and `next_wday` for the wday? "wday" is a common term in many programming languages, sounds like a good name to me.

andela-oakinniranye commented:
Thanks for the feedback @fxn, I'd update the `CHANGELOG.md` ASAP. To clarify the @matthewd's thought, I should be re-implementing the current `next_day/prev_day` methods to take an optional_argument, yeah?
I subscribe to @jeremy's thought on the `next_weekday` being non-weekend days, however the concern I seem to have with the `next_wday` is that `wday` returns a `Fixnum`, it seems logical that `prev_wday/next_wday` also return a similar `Fixnum`, which wouldn't also make sense as it would seem the argument is returned from the method. I'd like to know if the expressive `next_day_by_wday(1)` work better? Or other opinions on this?

andela-oakinniranye commented:
@fxn @jeremy @matthewd any more comments/updates on this? Also with the associated issue. #25325

Commit:
Allow passing parameters to the next_day to specify the next day required

In the previous implementation the next_*days were implemented, this removes this implementation and rather passes the day as an argument to the next_day, to better fit with the Calculations API.

This implementation allows methods such as:
	today = Date.current # => Thur, 30 June 2016
	today.next_day(:friday) # => Fri, 1 July 2016

This also adds a new method for fetching the next *day with the wday, next_wday and prev_day

	today = Date.current # => Thur, 30 June 2016
	today.next_wday(5) # => Fri, 1 July 2026
	today.prev_wday(3) # => Wed, 29 June 2016

andela-oakinniranye commented:
I have made updates to the PR with updates to the `next_day`, `prev_day` and added the `next_wday` and `prev_wday`

kamipo commented:
Closing in favor of #26600 and #30620.
